packages/nextjs/src/config/withSentryConfig.ts

```diff
@@ -50,11 +52,51 @@ function getFinalConfigObject(
     // eslint-disable-next-line @typescript-eslint/no-var-requires
     const { constructWebpackConfigFunction } = require('./webpack');
     return {
-      ...userNextConfigObject,
-      webpack: constructWebpackConfigFunction(userNextConfigObject, userSentryWebpackPluginOptions, userSentryOptions),
+      ...incomingUserNextConfigObject,
+      webpack: constructWebpackConfigFunction(
+        incomingUserNextConfigObject,
+        userSentryWebpackPluginOptions,
+        userSentryOptions,
+      ),
     };
   }
 
   // At runtime, we just return the user's config untouched.
-  return userNextConfigObject;
+  return incomingUserNextConfigObject;
+}
+
+/**
+ * Injects rewrite rules into the Next.js config provided by the user to tunnel
+ * requests from the `tunnelPath` to Sentry.
+ */
+function setUpTunnelRewriteRules(userNextConfig: NextConfigObject, tunnelPath: string): void {
+  const originalRewrites = userNextConfig.rewrites;
+
+  // This function doesn't take any arguments at the time of writing but we future-proof
+  // here in case Next.js ever decides to pass some
+  userNextConfig.rewrites = async (...args: unknown[]) => {
+    const injectedRewrites = [
+      {
+        source: `${tunnelPath}/:orgid/:projectid`, // without trailing slash
+        destination: 'https://:orgid.ingest.sentry.io/api/:projectid/envelope/',
+      },
+      {
+        source: `${tunnelPath}/:orgid/:projectid/`, // with trailing slash
+        destination: 'https://:orgid.ingest.sentry.io/api/:projectid/envelope/',
+      },
+    ];
+
+    if (typeof originalRewrites === 'function') {
+      // @ts-ignore weird args error
+      const intermediaryValue = await originalRewrites(...args);
+      if (Array.isArray(intermediaryValue)) {
+        intermediaryValue.push(...injectedRewrites);
+      } else {
+        intermediaryValue.beforeFiles.push(...injectedRewrites);
+      }
+      return intermediaryValue;
+    } else {
+      return injectedRewrites;
+    }
```

Comment:
L: What would you think of moving this logical branch first, so that the overall flow read more like:

```js
const injectedRewrites = [ ... ];

if (typeof originalRewrites !== 'function') {
  return injectedRewrites;
}

const mergedRewrites = await originalRewrites(...args);

if (Array.isArray(intermediaryValue)) {
  mergedRewrites.push(...injectedRewrites);
} else {
  mergedRewrites.beforeFiles.push(...injectedRewrites);
}

return mergedRewrites;
```
?

That way, the reader doesn't need to try to parse which ifs and elses are at which indentation level to figure out how they're grouped.